From 71e475249efe21125ba658f67e67f9258ccd9a02 Mon Sep 17 00:00:00 2001
From: Michel JAOUEN <michel.jaouen@st.com>
Date: Wed, 15 Oct 2025 11:00:06 +0200
Subject: [PATCH 930/954] feat(stm32mp2): fix shmem channel status read access

Use mmio_read_32 api to read channel_status to support polling, and add a
memory barrier to ensure shmem consistency before ringing doorbell.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git bd1475109d6e3c0d9f930feb398c818455d9955f feat(stm32mp2): fix shmem channel status read access]
Change-Id: I6cb6ad688d500219c76cd997a6385bd3d2bf2f8d
Signed-off-by: Michel JAOUEN <michel.jaouen@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/502086
Domain-Review: Maxime MERE <maxime.mere@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 plat/st/stm32mp2/scmi/scmi_common.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/plat/st/stm32mp2/scmi/scmi_common.c b/plat/st/stm32mp2/scmi/scmi_common.c
index c9f340e939..22f4f47c24 100644
--- a/plat/st/stm32mp2/scmi/scmi_common.c
+++ b/plat/st/stm32mp2/scmi/scmi_common.c
@@ -106,6 +106,14 @@ bool scmi_sys_pwr_state_set_supported;
  */
 static void scmi_ring_doorbell(void)
 {
+
+	/*
+	 * Ensure that any write to the SCMI payload area is seen by SCP before
+	 * we write to the doorbell register. If these writes were reordered
+	 * by the CPU then SCP would read stale payload data
+	 */
+	dmbst();
+
 	mmio_write_32(IPCC1_C1SCR, BIT_32(IPCC_CHANNEL + 16U));
 }
 
@@ -125,7 +133,7 @@ void scmi_channel_clear(void)
  */
 bool scmi_channel_busy(void)
 {
-	return (shmem->channel_status & SCMI_CHANNEL_STATUS_FREE) == 0;
+	return (mmio_read_32((uintptr_t)&shmem->channel_status) & SCMI_CHANNEL_STATUS_FREE) == 0;
 }
 
 /**
@@ -134,7 +142,7 @@ bool scmi_channel_busy(void)
  */
 bool scmi_channel_error(void)
 {
-	return (shmem->channel_status & SCMI_CHANNEL_STATUS_ERROR) != 0;
+	return (mmio_read_32((uintptr_t)&shmem->channel_status) & SCMI_CHANNEL_STATUS_ERROR) != 0;
 }
 
 /**
-- 
2.53.0

